File: src/cpp/fruchterman.cpp
#include <iostream>
#include <cmath>
#include "fruchterman.h"

void calRepulsive(const Node *nodes, const size_t nodeSize, Point *displacements, const float k2)
{
  for (size_t i = 0; i < nodeSize; i++) {
    for (size_t j = 0; j < nodeSize; j++) {
      if (i == j) {
        continue;
      }
      Node v = nodes[i];
      Node u = nodes[j];
      float vecX = v.x - u.x;
      float vecY = v.y - u.y;
      float vecLengthSqr = vecX * vecX + vecY * vecY;
      if (vecLengthSqr < 1e-6) {
        vecLengthSqr = 1.0;
        float sign = i > j ? 1.f : -1.f;
        vecX = sign * 0.01f;
        vecY = sign * 0.01f;
      }
      float common = k2 / vecLengthSqr;
      const float ret_x = vecX * common;
      const float ret_y = vecY * common;

      displacements[i].x += ret_x;
      displacements[i].y += ret_y;
    }
  }
}

void calAttractive(const Node *nodes, const size_t nodeSize, const Edge *edges, const size_t edgeSize, Point *displacements, const float k)
{
  for (size_t i = 0; i < edgeSize; i++) {
    Edge edge = edges[i];
    Node u = nodes[edge.sourceNodeArrayIdx];
    Node v = nodes[edge.targetNodeArrayIdx];

    if (edge.sourceNodeArrayIdx == edge.targetNodeArrayIdx) {
      continue;
    }

    float vecX = v.x - u.x;
    float vecY = v.y - u.y;
    float vecLength = sqrt(vecX * vecX + vecY * vecY);
    float common = (vecLength * vecLength) / k;
    displacements[edge.targetNodeArrayIdx].x -= (vecX / vecLength) / common;
    displacements[edge.targetNodeArrayIdx].y -= (vecY / vecLength) / common;
    displacements[edge.sourceNodeArrayIdx].x += (vecX / vecLength) * common;
    displacements[edge.targetNodeArrayIdx].y += (vecY / vecLength) * common;
  }
}

void calCluster(const Node *nodes, const size_t nodeSize, const Cluster *clusters, const size_t clusterSize, Point *displacements, const float clusterGravity, const float k)
{
  for (size_t i = 0; i < clusterSize; i++) {
    Cluster c = clusters[i];
    for (size_t j = 0; j < c.nodeSize; j++) {
      Node n = nodes[c.nodeArrayIdx[j]];
      float distLength = sqrt((n.x - c.cx) * (n.x - c.cx) + (n.y - c.cy) * (n.y - c.cy));
      float gravityForce = k * clusterGravity;

      displacements[j].x -= (gravityForce * (n.x - c.cx)) / distLength;
      displacements[j].y -= (gravityForce * (n.y - c.cy)) / distLength;
    }
  }
}

void calGravity(Node *nodes, size_t nodeSize, Point *displacements, float gravity, float k, Center *center)
{
  for (size_t i = 0; i < nodeSize; i++) {
    Node n = nodes[i];
    float gravityForce = 0.01f * k * gravity;

    displacements[i].x -= gravityForce * (n.x - center->x);
    displacements[i].y -= gravityForce * (n.y - center->y);
  }
}

File: src/fruchterman.ts
// const SPEED_DIVISOR = 800;

import { Edge, Node, Point } from "./types";

export function layout(nodes: Node[], edges: Edge[], displacements: Point[], k: number, k2: number) {
  calRepulsive(nodes, displacements, k2);
  if (edges) calAttractive(nodes, edges, displacements, k);
}

export function calRepulsive(nodes: Node[], displacements: Point[], k2: number) {
  nodes.forEach((v, i) => {
    displacements[i] = { x: 0, y: 0 };
    nodes.forEach((u, j) => {
      if (i === j) {
        return;
      }
      if (
        !isNumber(v.x) ||
        !isNumber(u.x) ||
        !isNumber(v.y) ||
        !isNumber(u.y)
      ) {
        return;
      }
      let vecX = v.x - u.x;
      let vecY = v.y - u.y;
      let vecLengthSqr = vecX * vecX + vecY * vecY;
      if (vecLengthSqr === 0) {
        vecLengthSqr = 1;
        const sign = i > j ? 1 : -1;
        vecX = 0.01 * sign;
        vecY = 0.01 * sign;
      }
      const common = k2 / vecLengthSqr;
      displacements[i].x += vecX * common;
      displacements[i].y += vecY * common;
    });
  });
  return displacements;
}

export function calAttractive(nodes: Node[], edges: Edge[], displacements: Point[], k: number) {
  edges.forEach((e) => {

    const uIndex = e.sourceNodeArrayIdx;
    const vIndex = e.targetNodeArrayIdx;
    if (!uIndex || !vIndex) return;
    if (uIndex === vIndex) {
      return;
    }
    const u = nodes[uIndex];
    const v = nodes[vIndex];
    if (!isNumber(v.x) || !isNumber(u.x) || !isNumber(v.y) || !isNumber(u.y)) {
      return;
    }
    const vecX = v.x - u.x;
    const vecY = v.y - u.y;
    const vecLength = Math.sqrt(vecX * vecX + vecY * vecY);
    const common = (vecLength * vecLength) / k;
    displacements[vIndex].x -= (vecX / vecLength) * common;
    displacements[vIndex].y -= (vecY / vecLength) * common;
    displacements[uIndex].x += (vecX / vecLength) * common;
    displacements[uIndex].y += (vecY / vecLength) * common;
  });
}

function isNumber(value: any) {
  return typeof value === 'number' && !isNaN(value);
}
